src/kernel/serial.zig

```diff
@@ -1,217 +1,41 @@
-}
-
-///
-/// Checks if the transmission buffer is empty, which means data can be sent.
-///
-/// Arguments:
-///     port: Port - The port to check.
-///
-/// Return: bool
-///     If the transmission buffer is empty.
-///
-fn transmitIsEmpty(port: Port) bool {
-    return arch.inb(@enumToInt(port) + 5) & 0x20 > 0;
-}
-
-///
-/// Write a byte to a serial port. Waits until the transmission queue is empty.
-///
-/// Arguments:
-///     char: u8 - The byte to send.
-///     port: Port - The port to send the byte to.
-///
-pub fn write(char: u8, port: Port) void {
-    while (!transmitIsEmpty(port)) {
-        arch.halt();
-    }
-    arch.outb(@enumToInt(port), char);
-}
-
-///
-/// Write a slice of bytes to a serial port. See write for more detailed information.
-///
-/// Arguments:
-///     str: []const u8 - The bytes to send.
-///     port: Port - The port to send the bytes to.
-///
-pub fn writeBytes(str: []const u8, port: Port) void {
-    for (str) |char| {
-        write(char, port);
-    }
-}
-
-///
-/// Initialise a serial port to a certain baudrate
-///
-/// Arguments
-///  IN baud: u32 - The baudrate to use. Cannot be more than MAX_BAUDRATE
-///  IN port: Port - The port to initialise
-///
-/// Error: SerialError
-///     InvalidBaudRate - The baudrate is 0 or greater than BAUD_MAX.
-///
-pub fn init(baud: u32, port: Port) SerialError!void {
-    // The baudrate is sent as a divisor of the max baud rate
-    const divisor: u16 = try baudDivisor(baud);
-    const port_int = @enumToInt(port);
-    // Send a byte to start setting the baudrate
-    arch.outb(port_int + LCR, lcrValue(0, false, false, 1) catch |e| {
-        panic(@errorReturnTrace(), "Failed to initialise serial output setup: {}", .{e});
-    });
-    // Send the divisor's lsb
-    arch.outb(port_int, @truncate(u8, divisor));
-    // Send the divisor's msb
-    arch.outb(port_int + 1, @truncate(u8, divisor >> 8));
-    // Send the properties to use
-    arch.outb(port_int + LCR, lcrValue(CHAR_LEN, SINGLE_STOP_BIT, PARITY_BIT, 0) catch |e| {
-        panic(@errorReturnTrace(), "Failed to setup serial properties: {}", .{e});
-    });
-    // Stop initialisation
-    arch.outb(port_int + 1, 0);
-
-    if (build_options.rt_test) runtimeTests();
-}
-
-test "lcrValue computes the correct value" {
-    // Check valid combinations
-    inline for ([_]u8{ 0, 5, 6, 7, 8 }) |char_len| {
-        inline for ([_]bool{ true, false }) |stop_bit| {
-            inline for ([_]bool{ true, false }) |parity_bit| {
-                inline for ([_]u1{ 0, 1 }) |msb| {
-                    const val = try lcrValue(char_len, stop_bit, parity_bit, msb);
-                    const expected = char_len & 0x3 |
-                        @boolToInt(stop_bit) << 2 |
-                        @boolToInt(parity_bit) << 3 |
-                        @intCast(u8, msb) << 7;
-                    testing.expectEqual(val, expected);
-                }
-            }
+pub const Serial = struct {
```

Comment:
Would need to import the x86 version of the serial via the arch. So would have to expose the serial in the arch file. Or could wait until the arch init and init the serial at the beginning of the arch.init() so wouldn't need to `pub` the serial in the arch file. But that would break logging for the memory managers. :( What do you think about this refactoring. Or could just keep it as is :P

Reply:
I think it's best as it is :) The way it is, the common code doesn't need to have any knowledge of how serial works under the hood and just lets the architecture do its thing, much like how allocators work.

Reply:
Unless you can think of a better way!